src/app/ReadClient.cpp

```diff
@@ -914,48 +949,92 @@ CHIP_ERROR ReadClient::SendSubscribeRequestImpl(const ReadPrepareParams & aReadP
-    _this->mNumRetries++;
+    VerifyOrReturnError(IsIdle(), CHIP_ERROR_INCORRECT_STATE);
+
+    auto timeTillNextResubscription = ComputeTimeTillNextSubscription();
+    ChipLogProgress(DataManagement,
+                    "Will try to resubscribe to %02x:" ChipLogFormatX64 " at retry index %" PRIu32 " after %" PRIu32
+                    "ms due to error %" CHIP_ERROR_FORMAT,
+                    GetFabricIndex(), ChipLogValueX64(GetPeerNodeId()), mNumRetries, timeTillNextResubscription,
+                    aTerminationCause.Format());
+    ReturnErrorOnFailure(ScheduleResubscription(timeTillNextResubscription, NullOptional, aTerminationCause == CHIP_ERROR_TIMEOUT));
+    return CHIP_NO_ERROR;
 }
 
-bool ReadClient::ResubscribeIfNeeded(uint32_t & aNextResubscribeIntervalMsec)
+void ReadClient::HandleDeviceConnected(void * context, OperationalDeviceProxy * device)
 {
-    bool shouldResubscribe       = true;
-    uint32_t intervalMsec        = 0;
-    aNextResubscribeIntervalMsec = 0;
-    if (mReadPrepareParams.mResubscribePolicy == nullptr)
+    ReadClient * const _this = static_cast<ReadClient *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    _this->mReadPrepareParams.mSessionHolder.Grab(device->GetSecureSession().Value());
+
+    auto err = _this->SendSubscribeRequest(_this->mReadPrepareParams);
+    if (err != CHIP_NO_ERROR)
     {
-        ChipLogDetail(DataManagement, "mResubscribePolicy is null");
-        return false;
+        _this->Close(err);
     }
-    mReadPrepareParams.mResubscribePolicy(mNumRetries, intervalMsec, shouldResubscribe);
-    if (!shouldResubscribe)
+}
+
+void ReadClient::HandleDeviceConnectionFailure(void * context, const ScopedNodeId & peerId, CHIP_ERROR err)
+{
+    ReadClient * const _this = static_cast<ReadClient *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    ChipLogError(DataManagement, "Failed to establish CASE for re-subscription with error '%" CHIP_ERROR_FORMAT "'", err.Format());
+
+    auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+    VerifyOrDie(caseSessionManager != nullptr);
+
+    _this->Close(err);
+}
+
+void ReadClient::OnResubscribeTimerCallback(System::Layer * apSystemLayer, void * apAppState)
+{
+    ReadClient * const _this = static_cast<ReadClient *>(apAppState);
+    VerifyOrDie(_this != nullptr);
+
+    CHIP_ERROR err;
+
+    ChipLogProgress(DataManagement, "OnResubscribeTimerCallback: DoCASE = %d", _this->mDoCaseOnNextResub);
+    _this->mNumRetries++;
+
+    if (_this->mDoCaseOnNextResub)
     {
-        ChipLogProgress(DataManagement, "Resubscribe has been stopped");
-        return false;
+        auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+        VerifyOrExit(caseSessionManager != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+        auto proxy = caseSessionManager->FindExistingSession(_this->mPeer);
+        if (proxy != nullptr)
+        {
+            proxy->Disconnect();
```

Comment:
> So the issue here is that just because we timed out in the sense of not getting a report that does not mean that we had an exchange-level timeout and so the session may not be marked defunct, I guess?

That's right.

Now that you've raised attention to it, I'm going to actually fix this logic to directly do this instead:
```
        if (_this->mReadPrepareParams.mSessionHolder) {
            _this->mReadPrepareParams.mSessionHolder.Get().Value()->AsSecureSession()->MarkAsDefunct();
        }
```

That should achieve an identical effect.

Importantly, with @tehampson 's changes coming up, methods like `FindExistingSession` will no longer be there nor make sense.

Reply:
Alas, this won't work until @tehampson 's changes are merged. Because right now, marking a session defunct has no effect on other existing session holders. Today, `OperationalDeviceProxy` instances just linger around, and so calling `FindOrEstablishSession` just returns back the existing object bound to the existing session. And that goes nowhere.

Reply:
Yep, that's right.  Calling Disconnect is the right thing for now.